### ISSUE TITLE
Avoid repeating target-independent tests

### ISSUE THREAD

**Issue by alexreinking** [halide/build_bot#126]
We currently run many JIT / correctness tests for GPU targets even when those tests do not have GPU schedules. For instance, it does not make sense to run `correctness_custom_allocator` once per target since it's CPU-only. There are very many tests like this.

**Comment by abadams** [comment]
It probably wouldn't be very hard to just make more groups in test/correctness/CMakeLists.txt instead of just "correctness". E.g. we could have a CPU group, and GPU group (things that skip if no gpu target is enabled) and a both group. Classifying tests into groups initially could probably be done with a grep invocation. When new tests are added the human that adds them can decide which group it should go in.

**Comment by alexreinking** [comment]
I agree: this could be done at the CTest/CMake level fairly easily (ie. no source changes). It would just be a touch tedious to validate that everything was sorted correctly.

**Comment by abadams** [comment]
As a first cut, anything without the string "gpu" in it should not be run when testing GPU backends, and anything without the string "hexagon" can be skipped when testing hexagon.
